config.org
#+title: Config
#+startup: overview
#+PROPERTY: header-args :results silent.
* Config
** Common
#+begin_quote
Here are some additional functions/macros that could help you configure Doom:

- `load!' for loading external *.el files relative to this one
- `use-package!' for configuring packages
- `after!' for running code after a package has loaded
- `add-load-path!' for adding directories to the `load-path', relative to
  this file. Emacs searches the `load-path' when you load packages with
  `require' or `use-package'.
- `map!' for binding new keys
#+end_quote

Load custom code
#+begin_src emacs-lisp
(add-to-list 'load-path "~/.doom.d/utils/")
#+end_src
#+begin_src elisp
(setq user-full-name "artur sharipov"
      user-mail-address "theartua@gmail.com")
(setq doom-localleader-key ",")
(setq mac-right-option-modifier 'meta)
#+end_src

Russian keyboard
#+begin_src elisp :tangle no
(use-package reverse-im
  :custom
  (reverse-im-input-methods '("russian-computer"))
  :config
  (reverse-im-mode t))
#+end_src

Initialize ~path~
#+begin_src emacs-lisp
(exec-path-from-shell-initialize)
#+end_src

Line numbers
#+begin_src  emacs-lisp
(setq display-line-numbers-type nil)
(global-display-line-numbers-mode 0)
#+end_src
Repeat mode
#+begin_quote
Repeat Mode lets you call multiple commands in a keymap without repeating the prefix each time. So you can press

C-x o, o, o, … to switch windows repeatedly. O will cycle backwards.
C-x u, u, u, … to undo repeatedly,
C-x @ C-n, n, p, f, b,… to traverse headings,
M-g n, n, p, n,… to navigate errors or grep matches, and so on.
#+end_quote
#+begin_src emacs-lisp
(repeat-mode t)
#+end_src

** FZF
#+begin_src emacs-lisp :tangle no
(map! :leader
      :desc "Commands"
      "SPC" 'fzf-projectile)
#+end_src
** Clojure
Cider buffer
#+begin_src emacs-lisp
(setq cider-save-file-on-load t)

(after! cider
  (setq cider-repl-pop-to-buffer-on-connect nil)
  (set-popup-rule! "^\\*cider-repl*" :size 0.4 :side 'right :select t :quit nil :ttl nil)
  (set-popup-rule! "^\\*cider-error*" :size 0.4 :side 'bottom :select t :quit t)
  (setq cljr-assume-language-context 'clj)
  ;; (setq cider-eldoc-display-for-symbol-at-point nil) ; disable cider showing eldoc during symbol at point

  ;; (setq clojure-indent-style 'align-arguments)
  ;; (setq clojure-align-forms-automatically nil)
  )
#+end_src
Debug macros =persistent-scope=
#+begin_src emacs-lisp
;; clj debug
(defun clj-insert-persist-scope-macro ()
  (interactive)
  (insert
   "(defmacro persist-scope
              \"Takes local scope vars and defines them in the global scope. Useful for RDD\"
              []
              `(do ~@(map (fn [v] `(def ~v ~v))
                  (keys (cond-> &env (contains? &env :locals) :locals)))))"))

(defun persist-scope ()
  (interactive)
  (let ((beg (point)))
    (clj-insert-persist-scope-macro)
    (cider-eval-region beg (point))
    (delete-region beg (point))
    (insert "(persist-scope)")
    (cider-eval-defun-at-point)
    (delete-region beg (point))))
#+end_src
** Appearance
#+begin_quote

Doom exposes five (optional) variables for controlling fonts in Doom. Here
are the three important ones:

size size ; + `doom-font'
+ `doom-variable-pitch-font'
+ `doom-big-font' -- used for `doom-big-font-mode'; use this for
  presentations or streaming.

They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
font string. You generally only need these two:
(setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
      doom-variable-pitch-font (font-spec :family "sans" :size 13))

There are two ways to load a theme. Both assume the theme is installed and
available. You can either set `doom-theme' or manually load a theme with the
`load-theme' function. This is the default:

#+end_quote

#+begin_src emacs-lisp
(setq doom-theme 'doom-one)

(setq doom-font (
font-spec :family "iosevka light"
:size 14))
#+end_src
Spellchecker highlight color
#+begin_src emacs-lisp :tangle no
(set-face-attribute 'flyspell-incorrect nil :underline '(:color "#ffc800" :style wave))
#+end_src
** Email client
#+begin_src elisp
(add-to-list 'load-path "/opt/homebrew/Cellar/mu/1.6.10/share/emacs/site-lisp/mu/mu4e")

(setq
 mue4e-headers-skip-duplicates t
 mu4e-view-show-images t
 mu4e-view-show-addresses t
 mu4e-use-fancy-chars t
 mu4e-compose-format-flowed nil
 mu4e-date-format "%y/%m/%d"
 mu4e-headers-date-format "%Y/%m/%d"
 mu4e-change-filenames-when-moving t)

(require 'smtpmail)
(setq message-send-mail-function 'smtpmail-send-it
   starttls-use-gnutls t
   smtpmail-starttls-credentials '(("smtp.gmail.com" 587 nil nil))
   smtpmail-auth-credentials
     '(("smtp.gmail.com" 587 "USERNAME@gmail.com" nil))
   smtpmail-default-smtp-server "smtp.gmail.com"
   smtpmail-smtp-server "smtp.gmail.com"
   smtpmail-smtp-service 587)

(set-email-account! "theartua"
  '((smtpmail-smtp-user     . "theartua@gmail.com"))
  t)

(set-email-account! "hs"
  '((smtpmail-smtp-user     . "artur.sharipov@health-samurai.io")
    ;; (smtpmail-smtp-server "smtp.gmail.com")
    ;; (smtpmail-smtp-service 587)
    ) t)

(setq mu4e-context-policy 'ask-if-none
      mu4e-compose-context-policy 'always-ask)
#+end_src
** RSS reader
#+begin_src elisp
(setq rmh-elfeed-org-files '("/Users/artua/org/elfeed.org"))

(map! :leader
      (:prefix "o"
        :desc "Elfeed" "e" 'elfeed))

(map! (:map elfeed-search-mode-map
       :localleader
       :desc "Update feed"
       "u" #'elfeed-update))

(add-hook! 'elfeed-search-mode-hook 'elfeed-update)
#+end_src
** Evil
Friendly evil escape
#+begin_src emacs-lisp
(setq evil-move-cursor-back nil)
#+end_src
Disable evil clipboard
#+begin_src emacs-lisp
(setq x-select-enable-clipboard t)
#+end_src
** Autosave
#+begin_src elisp
(use-package super-save
  :config
  (add-to-list 'super-save-hook-triggers 'find-file-hook)
  (setq super-save-remote-files nil)
  (setq super-save-exclude '(".gpg", ".pyc", ".elc"))
  (setq super-save-auto-save-when-idle t)
  (setq auto-save-default nil)
  (super-save-mode +1))

#+end_src
** Lisp structure editing
#+begin_src elisp :tangle no
(use-package! evil-lisp-state
  :init
  (setq evil-lisp-state-global t)
  :config
  (map! :leader :desc "lisp" "k" evil-lisp-state-map))

(after! which-key
  (add-to-list
    'which-key-replacement-alist
    '((nil . "evil-lisp-state-") . (nil . "")))
#+end_src
** Expand region
#+begin_src elisp
(global-set-key (kbd "C-s") 'er/expand-region)
#+end_src
** Spacemacs like surround and go to test hotkey
#+begin_src elisp
(map! :v "s" #'evil-surround-region)

(map! :leader
      ;;"TAB" #'evil-switch-to-windows-last-buffer
      "p a" #'projectile-toggle-between-implementation-and-test)
#+end_src
** Helm
ff more useful
#+begin_src elisp
(map! (:after helm-files :map (helm-find-files-map helm-read-file-map)
       "C-h" #'helm-find-files-up-one-level
       "C-j" #'helm-next-line
       "C-k" #'helm-previous-line
       "C-l" #'helm-execute-persistent-action))

(map! (:after helm-files :map (helm-find-files-map helm-read-file-map)
       "C-b" #'helm-find-files-up-one-level
       "C-f" #'helm-execute-persistent-action))
#+end_src

** Vs code like navigation
#+begin_src elisp
(global-set-key (kbd "s-P") 'helm-M-x)
(global-set-key (kbd "s-p") 'helm-projectile)
(global-set-key (kbd "s-O") 'imenu)
(global-set-key (kbd "s-o") 'helm-find-files)
(global-set-key (kbd "s-b") '+neotree/open)
(global-set-key (kbd "s-w") 'kill-current-buffer)
(global-set-key (kbd "s-g") 'magit-status)
(global-set-key (kbd "s-.") 'eglot-code-actions)
#+end_src
Multiple cursors (evil depended)
#+begin_src emacs-lisp :tangle no
(global-set-key (kbd "s-d") 'evil-multiedit-match-and-next)
#+end_src
** Org Mode
#+begin_src emacs-lisp
(setq org-directory "~/org/")
#+end_src
Pomodoro osx notifications
#+begin_src elisp
(defun notify-osx (title message)
  (call-process "terminal-notifier"
                nil 0 nil
                "-group" "Emacs"
                "-title" title
                "-sender" "org.gnu.Emacs"
                "-message" message))

(add-hook 'org-pomodoro-finished-hook
          (lambda ()
          (notify-osx "Pomodoro completed!" "Time for a break.")))

(add-hook 'org-pomodoro-break-finished-hook
          (lambda ()
          (notify-osx "Pomodoro Short Break Finished" "Ready for Another?")))

(add-hook 'org-pomodoro-long-break-finished-hook
          (lambda ()
          (notify-osx "Pomodoro Long Break Finished" "Ready for Another?")))

(add-hook 'org-pomodoro-killed-hook
          (lambda ()
          (notify-osx "Pomodoro Killed" "One does not simply kill a pomodoro!")))

#+end_src

Evaluate bash codeblocks directly to tmux session
#+begin_src elisp
(use-package ob-tmux
  ;; Install package automatically (optional)
  :ensure t
  :custom
  (org-babel-default-header-args:tmux
   '((:results . "silent")	;
     (:session . "default")	; The default tmux session to send code to
     (:socket  . nil)))		; The default tmux socket to communicate with
  ;; The tmux sessions are prefixed with the following string.
  ;; You can customize this if you like.
  (org-babel-tmux-session-prefix "ob-")
  ;; The terminal that will be used.
  ;; You can also customize the options passed to the terminal.
  ;; The default terminal is "gnome-terminal" with options "--".
  ;;
  ;; Directly to emacs window:
  ;; #!/bin/bash
  ;; emacsclient -e "(progn (vterm) (vterm-send-string \"$*\") (vterm-send-return))"
  (org-babel-tmux-terminal "/Applications/Alacritty.app/Contents/MacOS/alacritty")
  (org-babel-tmux-terminal-opts '("-t" "ob-tmux" "-e")))
#+end_src
** Obsidian
#+begin_src emacs-lisp
(use-package obsidian
  :ensure t
  :demand t
  :config
  (obsidian-specify-path "~/vault")
  (global-obsidian-mode t)
  ;; :custom
  ;; This directory will be used for `obsidian-capture' if set.
  ;; (obsidian-inbox-directory "Inbox")
  :bind (:map obsidian-mode-map
  ;; Replace C-c C-o with Obsidian.el's implementation. It's ok to use another key binding.
  ("C-c C-o" . obsidian-follow-link-at-point)
  ;; Jump to backlinks
  ("C-c C-b" . obsidian-backlink-jump)
  ;; If you prefer you can use `obsidian-insert-link'
  ("C-c C-l" . obsidian-insert-wikilink)))
#+end_src
** Non evil
#+begin_src emacs-lisp
(global-set-key (kbd "C-:") 'avy-goto-char)
(global-set-key (kbd "C-h") 'delete-backward-char)
(global-set-key (kbd "C-.") 'repeat)
#+end_src
Wrapping
Now try selecting a region and press any of the following keys: ", ', (, {, [.
#+begin_src emacs-lisp
(use-package! wrap-region
  :init
  (wrap-region-mode t))
#+end_src
Go to last change like vim
#+begin_src emacs-lisp
(require 'goto-last-change)
(map! :leader "." 'goto-last-change)
#+end_src
Multiple cursors
#+begin_src emacs-lisp
(global-set-key (kbd "s-d") 'mc/mark-next-like-this-word)
(global-set-key (kbd "s-D") 'mc/mark-previous-like-this-word)
#+end_src
Emacs "Visual" mode
 ~C-SPC~ is default
#+begin_src emacs-lisp
(global-set-key (kbd "M-SPC") 'set-mark-command)
#+end_src


config.el
(setq user-full-name "artur sharipov"
      user-mail-address "theartua@gmail.com")

(setq doom-localleader-key ",")


(use-package reverse-im
  :custom
  (reverse-im-input-methods '("russian-computer"))
  :config
  (reverse-im-mode t))


(setq mac-right-option-modifier 'control)

(exec-path-from-shell-initialize)

(setq display-line-numbers-type nil)
(global-display-line-numbers-mode 0)

(map! :leader
      :desc "Commands"
      "SPC" 'fzf-projectile)

(setq cider-save-file-on-load t)

(after! cider
  (setq cider-repl-pop-to-buffer-on-connect nil)
  (set-popup-rule! "^\\*cider-repl*" :size 0.4 :side 'right :select t :quit nil :ttl nil)
  (set-popup-rule! "^\\*cider-error*" :size 0.4 :side 'bottom :select t :quit t)
  (setq cljr-assume-language-context 'clj)
  ;; (setq cider-eldoc-display-for-symbol-at-point nil) ; disable cider showing eldoc during symbol at point

  ;; (setq clojure-indent-style 'align-arguments)
  ;; (setq clojure-align-forms-automatically nil)
  )

;; clj debug
(defun clj-insert-persist-scope-macro ()
  (interactive)
  (insert
   "(defmacro persist-scope
              \"Takes local scope vars and defines them in the global scope. Useful for RDD\"
              []
              `(do ~@(map (fn [v] `(def ~v ~v))
                  (keys (cond-> &env (contains? &env :locals) :locals)))))"))

(defun persist-scope ()
  (interactive)
  (let ((beg (point)))
    (clj-insert-persist-scope-macro)
    (cider-eval-region beg (point))
    (delete-region beg (point))
    (insert "(persist-scope)")
    (cider-eval-defun-at-point)
    (delete-region beg (point))))

(setq doom-theme 'doom-one)

(setq doom-font (
font-spec :family "iosevka"
:size 14))

(set-face-attribute 'flyspell-incorrect nil :underline '(:color "#ffc800" :style wave))

(add-to-list 'load-path "/opt/homebrew/Cellar/mu/1.6.10/share/emacs/site-lisp/mu/mu4e")

(setq
 mue4e-headers-skip-duplicates t
 mu4e-view-show-images t
 mu4e-view-show-addresses t
 mu4e-use-fancy-chars t
 mu4e-compose-format-flowed nil
 mu4e-date-format "%y/%m/%d"
 mu4e-headers-date-format "%Y/%m/%d"
 mu4e-change-filenames-when-moving t)

(require 'smtpmail)
(setq message-send-mail-function 'smtpmail-send-it
   starttls-use-gnutls t
   smtpmail-starttls-credentials '(("smtp.gmail.com" 587 nil nil))
   smtpmail-auth-credentials
     '(("smtp.gmail.com" 587 "USERNAME@gmail.com" nil))
   smtpmail-default-smtp-server "smtp.gmail.com"
   smtpmail-smtp-server "smtp.gmail.com"
   smtpmail-smtp-service 587)

(set-email-account! "theartua"
  '((smtpmail-smtp-user     . "theartua@gmail.com"))
  t)

(set-email-account! "hs"
  '((smtpmail-smtp-user     . "artur.sharipov@health-samurai.io")
    ;; (smtpmail-smtp-server "smtp.gmail.com")
    ;; (smtpmail-smtp-service 587)
    ) t)

(setq mu4e-context-policy 'ask-if-none
      mu4e-compose-context-policy 'always-ask)

(setq rmh-elfeed-org-files '("/Users/artua/org/elfeed.org"))

(map! :leader
      (:prefix "o"
        :desc "Elfeed" "e" 'elfeed))

(map! (:map elfeed-search-mode-map
       :localleader
       :desc "Update feed"
       "u" #'elfeed-update))

(add-hook! 'elfeed-search-mode-hook 'elfeed-update)

(setq evil-move-cursor-back nil)

(setq x-select-enable-clipboard t)

(use-package super-save
  :config
  (add-to-list 'super-save-hook-triggers 'find-file-hook)
  (setq super-save-remote-files nil)
  (setq super-save-exclude '(".gpg", ".pyc", ".elc"))
  (setq super-save-auto-save-when-idle t)
  (setq auto-save-default nil)
  (super-save-mode +1))

(use-package! evil-lisp-state
  :init
  (setq evil-lisp-state-global t)
  :config
  (map! :leader :desc "lisp" "k" evil-lisp-state-map))

(after! which-key
  (add-to-list
    'which-key-replacement-alist
    '((nil . "evil-lisp-state-") . (nil . ""))))

(map! :v "s" #'evil-surround-region)

(map! :leader
      ;;"TAB" #'evil-switch-to-windows-last-buffer
      "p a" #'projectile-toggle-between-implementation-and-test)

(map! :n "C-w" #'er/expand-region
      :m "C-w" #'er/expand-region
      :i "C-w" #'er/expand-region
      :r "C-w" #'er/expand-region
      :v "C-w" #'er/expand-region)

(map! (:after helm-files :map (helm-find-files-map helm-read-file-map)
       "C-h" #'helm-find-files-up-one-level
       "C-j" #'helm-next-line
       "C-k" #'helm-previous-line
       "C-l" #'helm-execute-persistent-action))

(map! "s-P" #'helm-M-x
      "s-p" #'helm-projectile
      "s-o" #'helm-find-files
      "s-O" #'imenu
      "s-b" #'+neotree/open
      ;; "s-b" #'fzf-switch-buffer
      "s-w" #'kill-current-buffer
      "s-g" #'magit-status
      "s-d" #'evil-multiedit-match-and-next
      "s-." #'lsp-execute-code-action)

(setq org-directory "~/org/")

(defun notify-osx (title message)
  (call-process "terminal-notifier"
                nil 0 nil
                "-group" "Emacs"
                "-title" title
                "-sender" "org.gnu.Emacs"
                "-message" message))

(add-hook 'org-pomodoro-finished-hook
          (lambda ()
          (notify-osx "Pomodoro completed!" "Time for a break.")))

(add-hook 'org-pomodoro-break-finished-hook
          (lambda ()
          (notify-osx "Pomodoro Short Break Finished" "Ready for Another?")))

(add-hook 'org-pomodoro-long-break-finished-hook
          (lambda ()
          (notify-osx "Pomodoro Long Break Finished" "Ready for Another?")))

(add-hook 'org-pomodoro-killed-hook
          (lambda ()
          (notify-osx "Pomodoro Killed" "One does not simply kill a pomodoro!")))

(use-package ob-tmux
  ;; Install package automatically (optional)
  :ensure t
  :custom
  (org-babel-default-header-args:tmux
   '((:results . "silent")	;
     (:session . "default")	; The default tmux session to send code to
     (:socket  . nil)))		; The default tmux socket to communicate with
  ;; The tmux sessions are prefixed with the following string.
  ;; You can customize this if you like.
  (org-babel-tmux-session-prefix "ob-")
  ;; The terminal that will be used.
  ;; You can also customize the options passed to the terminal.
  ;; The default terminal is "gnome-terminal" with options "--".
  ;;
  ;; Directly to emacs window:
  ;; #!/bin/bash
  ;; emacsclient -e "(progn (vterm) (vterm-send-string \"$*\") (vterm-send-return))"
  (org-babel-tmux-terminal "/Applications/Alacritty.app/Contents/MacOS/alacritty")
  (org-babel-tmux-terminal-opts '("-t" "ob-tmux" "-e")))
